### PR TITLE
Optimize sequential insert into memtable - Part 1: Interface

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary: Currently our skip-list have an optimization to speedup sequential
inserts from a single stream, by remembering the last insert position.
We extend the idea to support sequential inserts from multiple streams,
and even tolerate small reordering wihtin each stream.

This PR is the interface part adding the following:
- Add `memtable_insert_prefix_extractor` to allow specifying prefix for each key.
- Add `InsertWithHint()` interface to memtable, to allow underlying
  implementation to return a hint of insert position, which can be later
  pass back to optimize inserts.
- Memtable will maintain a map from prefix to hints and pass the hint
  via `InsertWithHint()` if `memtable_insert_prefix_extractor` is non-null.

Test Plan: `make all check`. Will add more test and benchmarks in later
PR.

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
cc @mvm3k @al13n321

File: include/rocksdb/options.h [hunk 1]
```diff
@@ -405,6 +405,13 @@ struct ColumnFamilyOptions {
   // Default: nullptr
   std::shared_ptr<const SliceTransform> prefix_extractor;
 
+  // If non-nullptr, memtable will use the specified function to extract
+  // prefixes for keys, and optimize CPU usages for sequential inserts within
+  // each prefix. Currently only affects the default skip-list-based memtable. 
+  //
+  // Default: nullptr
+  std::shared_ptr<const SliceTransform> memtable_insert_prefix_extractor;
```
Comment (siying) [inline include/rocksdb/options.h]: Can we include "hint" in the option name?

File: include/rocksdb/options.h [hunk 2]
```diff
@@ -405,6 +405,13 @@ struct ColumnFamilyOptions {
   // Default: nullptr
   std::shared_ptr<const SliceTransform> prefix_extractor;
 
+  // If non-nullptr, memtable will use the specified function to extract
+  // prefixes for keys, and optimize CPU usages for sequential inserts within
+  // each prefix. Currently only affects the default skip-list-based memtable.
```
Comment (siying) [inline include/rocksdb/options.h]: Should introduce in what use cases users can consider to use this feature and how should they determine the prefix.

File: db/memtable.h [hunk 3]
```diff
@@ -387,6 +389,12 @@ class MemTable {
 
   Env* env_;
 
+  // Extract sequential insert prefixes.
+  const SliceTransform* insert_prefix_extractor_;
+
+  // Insert hints for each prefix.
+  std::unordered_map<Slice, MemTableInsertHint, SliceHasher> insert_hints_;
```
Comment (al13n321) [inline db/memtable.h]: Would be nice to account for the memory used by this map in WriteBufferManager and MemTable::ApproximateMemoryUsage(). Maybe it's possible to use the same ConcurrentArena allocator for this map as for the memtable contents (with a no-op deallocate(), with an assumption that unordered_map doesn't deallocate much); that would automatically count the memory usage in both places, and may also improve performance.

Comment (yiwu-arbug) [inline db/memtable.h]: I'm using the same Arena object passing to InlineSkipList to allocate the hitns. My understanding is the memory usage will be counted in this way, but I'll double check.

File: include/rocksdb/options.h (hunk 2 above)
Comment (yiwu-arbug) [inline include/rocksdb/options.h]: Sure. Will update soon.

**Comment by yiwu-arbug** [comment]
PR updated:
 - Improve inline comment. Change the option name to `memtable_insert_with_hint_prefix_extractor`.
 - whitelist keys from using hints if prefix for the key is empty.
 - disabling the feature if concurrent writes are allowed.

Pending work:
 - Unit test

Will appear in another PR:
  - Make `Arena` compatible with `std::allocator` and use it for allocating the hints map.

**Comment by siying** [comment]
Many tests are failing.

**Comment by yiwu-arbug** [comment]
Fix test.

**Comment by yiwu-arbug** [comment]
Adding db_memtable_test to test the new options. Build will fail before landing #1449.

File: db/memtable.cc [hunk 4]
```diff
@@ -431,7 +434,16 @@ void MemTable::Add(SequenceNumber s, ValueType type,
   memcpy(p, value.data(), val_size);
   assert((unsigned)(p + val_size - buf) == (unsigned)encoded_len);
   if (!allow_concurrent) {
-    table->Insert(handle);
+    // Extract prefix for insert with hint.
+    Slice prefix;
+    if (insert_with_hint_prefix_extractor_ != nullptr) {
+      prefix = insert_with_hint_prefix_extractor_->Transform(key_slice);
+    }
+    if (prefix.empty()) {
```
Comment (yiwu-arbug) [inline db/memtable.cc]: I want to provide a way for user to whitelist keys if they don't use hint for some of the keys (e.g. people mix keys with prefix and without prefix in the same db). Plus it doesn't make sense if prefix is empty.

In the updated revision I also check `prefix_extractor.InDomain()` and whitelist keys not in domain.

**Comment by siying** [comment]
Also, mention it in HISTORY.md

**Comment by yiwu-arbug** [comment]
Rebase and update HISTORY.md.

**Comment by yiwu-arbug** [comment]
clang-format

**Comment by yiwu-arbug** [comment]
Fix `InsertWithHint` not mark with `override`.

**Pull request closed without merging**